winConsoleUIATextInfo: use rangeFromPoint instead of broken getVisibleRanges

michaelDCurran opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Fixes #10191 
Fixes #10119

### Summary of the issue:
In Windows consoles with UIA, NVDA currently makes use of IUIAutomationTextPattern::getVisibleRanges to find the bounds of what is visibly on screen. This is both for bounding the review cursor, and for collecting the on-screen text for diffing, in order to speak what has changed.
However, IUIAutomationTextPattern:getVisibleRanges can eventually start giving back junk text ranges that are positioned way above, and outside of the real visible bounds. This seems to start happening once a great deal of text has been scrolled up the screen.
This not only causes the review cursor to act strangely or get stuck when moving to the top or bottom, but it causes very old and invalid text to be spoken when things change on screen.
 
### Description of how this pull request fixes the issue:
Instead of getVisibleRanges, use IUIAutomationTextPattern::rangeFromPoint at the top left and bottom right of the console window.

### Testing performed:
Tested in both wsl (Ubuntu) and cmd.
Tested when older alpha builds of NVDA got into a state where the console no longer made sence with getVisibleRanges; this approach seemed to give back correct text and the review cursor no longer got stuck.

### Known issues with pull request:
Like the older legacy winConsole support, if moving to the bottom with the review cursor and then going up by line, if there are blank lines in the console, these are not skipped. this is a little annoying for efficiency, but it is the same behaviour we originally had for many years, and blank lines are certainly better than  invalid text or getting stuck.

### Change log entry:
None.

michaelDCurran committed:
WinConsoleUIATextInfo:  As IUIAutomationTextPattern::getVisibleRanges is very broken, use IUIAutomationTextPattern::rangeFromPoint for the top left and bottom right of the console window.

michaelDCurran commented:
@codeofdusk  I'd appreciate your help in testing this where you can. I was originally against this approach, but since getvisibleRanges is very broken,  I don't think we have a choice anymore.

michaelDCurran committed:
Fix linting issues.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -272,13 +280,6 @@ def _get_TextInfo(self):
 		bounds review to the visible text."""
 		return consoleUIATextInfo
 
-	def _getTextLines(self):
```
I assume this now works when inherited from the base class, and that's the reason why you removed this?

michaelDCurran replied:
This is correct. The base implementation of _getTextLines seems to be fine as far as I understand.

codeofdusk commented:
> @codeofdusk I'd appreciate your help in testing this where you can. I was originally against this approach, but since getvisibleRanges is very broken, I don't think we have a choice anymore.

After preliminary testing, I found that:

* The test cases from #9891 and #10406 do not result in broken autorread and a crashed console respectively.
* There is a slight autoread performance hit after the changes (it appears that `_getTextLines` runs a bit slower).
* When a console is flooded with many new text events, it is completely unusable – it is, of course, expected that autoread will stop functioning, but even text review breaks after these changes.

michaelDCurran commented:
I'm guessing that perhaps when review stops working when the console is 
flooded,  this is when in the existing code new text announced was 
invalid, and review also was broken? Or are you saying that this pr is 
actually worse?

I think I can definitly address the  speed issue by putting back 
_getTextLines that calls IUIAutomationTextRange::getText directly. The 
base _getTextLines breaks text by unit line which is rather slow for 
UIA. Rather as this UIA implementation contains linefeed characters, we 
can just do a python string splitlines.

codeofdusk commented:
> I'm guessing that perhaps when review stops working when the console is flooded,  this is when in the existing code new text announced was invalid, and review also was broken? Or are you saying that this pr is actually worse? I think I can definitly address the  speed issue by putting back _getTextLines that calls IUIAutomationTextRange::getText directly. The base _getTextLines breaks text by unit line which is rather slow for UIA. Rather as this UIA implementation contains linefeed characters, we can just do a python string splitlines.

This PR is worse. Previously, autoread stopped completely but review still worked.

Also, a `RuntimeError` is thrown when the console exits after this PR.

michaelDCurran committed:
WinConsoleUIA: override _getTextLines with a faster implementation that just uses Python string splitlines.

michaelDCurran committed:
Fix linting issues.

codeofdusk commented:
It seems that the `move` method still uses `isTextRangeOffscreen`. Is this intentional?

codeofdusk commented:
Also, it seems that your `POSITION_LAST` initialization assumes that `_expandCollapseBeforeReview` is `True`. Since #9802 was merged, there might not be a need to do this any more. Consider removing the `_expandCollapseBeforeReview` override.

michaelDCurran committed:
Don't use isTextRangeOffscreen, and further improve POSITION_LAST to fix review bounds.

michaelDCurran committed:
Fix linting issues.

michaelDCurran committed:
WinConsoleUIATextInfo: ensure ranges are collapsed after moves.

michaelDCurran committed:
Remove TextInfo._expandCollapseBeforeReview as WinConsoleUIA no longer needs it.

michaelDCurran committed:
Merge branch 'master' into i10119

michaelDCurran commented:
@codeofdusk  I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
I have entirely removed _expandCollapseBeforeReview as requested.
I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
If you could run through your tests again it would be much appreciated.
If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.

codeofdusk commented:
> @codeofdusk I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
> I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
> I have entirely removed _expandCollapseBeforeReview as requested.
> I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
> If you could run through your tests again it would be much appreciated.
> If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.

Things are a lot better, but:

STR:
1. Open a console.
2. Invoke "review bottom line".
3. Invoke "review current word".

Expected: the last word of the prompt is read out.
Actual: "blank" is reported.

codeofdusk commented:
> > @codeofdusk I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
> > I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
> > I have entirely removed _expandCollapseBeforeReview as requested.
> > I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
> > If you could run through your tests again it would be much appreciated.
> > If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.
> 
> Things are a lot better, but:
> 
> STR:
> 
> 1. Open a console.
> 2. Invoke "review bottom line".
> 3. Invoke "review current word".
> 
> Expected: the last word of the prompt is read out.
> Actual: "blank" is reported.

Nevermind, I think that was confusion on my part (as now consoles have blank lines at the end per known issues). The confusing part is that word and line movement have different bounds (i.e. I can access some parts of the console with word review that I can't with line review, and vice versa).

michaelDCurran committed:
WinConsoleUIATextInfo: when calculating word / line offsets, grab text directly with IUIAutomationTextRange::getText as  WinConsoleUIATextInfo.text has been changed to return a space for empty ranges.

michaelDCurran commented:
It seems moving by word was a little broken after your pr was merged to 
return a space from the text property. I have worked around this by 
calling IUIAutomationTextRange::getText directly from the word and line 
offset   calculation functions.

codeofdusk reviewed (approved):
Looks good to me!

codeofdusk commented:
Also, I re-tested the console flooding case. This PR doesn't regress behavior.

codeofdusk commented:
@michaelDCurran  When do you plan to merge this PR?

michaelDCurran commented:
Branch policy dictates that I require a review approval from a 
"collaborator" such as @leonardder before I can merge.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -18,26 +18,31 @@
+			# Therefore use the bottom left, then move the  to last character on that line.
+			tempInfo = self.__class__(obj, obj.location.bottomLeft)
+			tempInfo.expand(textInfos.UNIT_LINE)
+			UIATextInfo.move(tempInfo, textInfos.UNIT_CHARACTER, -1, endPoint="end")
```
What happens if we don't do this? We already expanded to line, so the endpoint should be the end of the line already?

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -54,14 +59,12 @@ def collapse(self, end=False):
-			visiLength = visiRanges.length
-			if visiLength > 0:
-				oldRange = self._rangeObj.clone()
+			boundingInfo = self.obj.makeTextInfo(textInfos.POSITION_ALL)
```
Consistency with other code

```suggestion
			boundingInfo = self.__class__(obj, textInfos.POSITION_ALL)
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -116,13 +119,24 @@ def move(self, unit, direction, endPoint=None):
-				oldRange
-				and isTextRangeOffscreen(self._rangeObj, visiRanges)
-				and not isTextRangeOffscreen(oldRange, visiRanges)
+				oldInfo
```
Could you add a small comment that explains what this if statement tries to do? It is hard to read at first glance.

LeonarddeR reviewed (changes requested):
Just some small things

michaelDCurran committed:
Address review actions.

michaelDCurran commented:
@leonardder I have improved comments as requested. However, I did not take your suggestion of changing the makeTextInfo call into a direct WinConsoleUIATextInfo construction as this is not needed in the move method -- it can stay high-level. The only reason I did it in WinConsoleUIATextInfo's __init__ method was because I was passing in _rangeObj which makeTextInfo does not allow.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -116,16 +123,32 @@ def move(self, unit, direction, endPoint=None):
+		# Console textRanges have access to the entire console buffer.
+		# However, we want to limit ourselves to onscreen text.
+		# Therefore, if the textInfo was originally visible,
+		# but we are now aboe or below the visible range,
```
Typo:

```suggestion
		# but we are now above or below the visible range,
```

michaelDCurran committed:
Update source/NVDAObjects/UIA/winConsoleUIA.py

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>
